src/proof/PGHelp.cc

```diff
@@ -90,31 +90,37 @@ void ProofGraph::getGraphInfo()
 }
 
 std::vector<clauseid_t> ProofGraph::topolSortingTopDown() const {
-    std::vector<clauseid_t> DFS;
-    std::deque<clauseid_t> q;
-    DFS.reserve(getGraphSize());
-    // Enqueue leaves first
-    q.assign(leaves_ids.begin(),leaves_ids.end());
-    do {
-        clauseid_t id = q.front();
-        q.pop_front();
+    std::vector<clauseid_t> order;
+    std::vector<clauseid_t> unprocessed;
+    std::vector<char> marks (getGraphSize(), 0);
+    order.reserve(getGraphSize());
+    unprocessed.push_back(getRoot()->getId());
+    while (not unprocessed.empty()) {
+        auto id = unprocessed.back();
+        auto & mark = marks.at(id);
+        if (mark == 2) { // already processed
+            unprocessed.pop_back();
+            continue;
+        }
         ProofNode * n = getNode(id);
         assert(n);
-        if (not isSetVisited1(id)) {
-            // Process node if both antecedents visited
-            if ((not n->getAnt1() or isSetVisited1(n->getAnt1()->getId())) and
-                (not n->getAnt2() or isSetVisited1(n->getAnt2()->getId()))) {
-                for (clauseid_t resolvent_id: n->getResolvents()) {
-                    if (getNode(resolvent_id)) q.push_back(resolvent_id);
-                }
-                setVisited1(id);
-                DFS.push_back(id);
-            }
+        if (mark == 1 or n->isLeaf()) { // returning from processing parents or is leaf => mark as done
+            mark = 2;
+            order.push_back(id);
+            unprocessed.pop_back();
+            continue;
         }
+        mark = 1;
+        auto processParent = [&](auto parentId) {
```

Comment:
This lambda is constructed in every loop cycle, but it sufficient to construct it only only before the loop, since it contains only references to the outer variables.

Reply:
Hoisted!